bm-timeline/src/Adoption.hs

```diff
@@ -0,0 +1,222 @@
+  let bp_diffusion1 n = foldl (\acc n2 -> bp_diffusion2 n n2 : acc) [] $ reverse nodes
+  let bp_diffusion = foldl (\acc n -> [["src node", "target node", "min","q1","median","q3","max"]] ++ bp_diffusion1 n ++ acc) [] $ reverse nodes
+  -- let bp_diff_headers = map (const ["src node", "tgt node", "min","q1","median","q3","max"]) nodes
+  withFile "adoption.csv" WriteMode $ \h ->
+      output_csv h $ named_columns $ [ (["slot", "timestamp", "node id", "tadopted", "num tx", "block size"], tuple6_in_columns(tadopted))
+                                     , (["node id", "min", "q1", "median", "q3", "max"], bp_adopted) ]
+                                    <> zip cdf_tb_headers cdf_tbetween
+  withFile "diffusion.csv" WriteMode $ \h ->
+      output_csv h $ named_columns [ (["slot", "src node", "tgt node", "tdiffusion", "num tx", "block size"], tuple6_in_columns(tdiffusion))
+                                   , (head bp_diffusion, tail bp_diffusion) ]
+                                    -- <> zip bp_diff_headers bp_diffusion
+
+  return ()
+
+
+calculate_time_between :: HMAdopted -> NodeId -> [Milliseconds]
+calculate_time_between hm n =
+  let adls = sortOn fst $ filter (\(_,(node,_,_,_)) -> node == n) $ HM.toList hm
+  in map timedifference $ zip adls (tail adls)
+ where
+   timedifference :: ((SlotNum,ValAdopted),(SlotNum,ValAdopted)) -> Milliseconds
+   timedifference ((_sn1,(_,ts1,_,_)), (_sn2,(_,ts2,_,_))) =
+     round $ 1000.0 * diffUTCTime ts2 ts1
+
+{-
+setAt :: Int -> a -> [a] -> [a]
+setAt _ _val [] = []
+setAt 0 val (_v : vs) = val : vs
+setAt idx val (v : vs) = v : setAt (idx - 1) val vs
+-}
+
+reconstruct :: [NodeId] -> [SlotNum] -> (HMChain, HMAdopted, HMLeader)
+            -> [AdoptedTimeTuple] -> [DiffusionTimeTuple]
+            -> IO ([AdoptedTimeTuple], [DiffusionTimeTuple])
+reconstruct _ [] _ acc1 acc2 = pure (acc1, acc2)
+reconstruct nodes (sn : sns) (hmch, hmad, hmld) acc1 acc2 = do
+  leader <- reconstruct_leader sn (HM.lookup sn hmld)
+  recadopt <- reconstruct_adopt sn leader (HM.lookup sn hmad)
+  (tdiffadopted, tdiffusion0) <- case recadopt of
+    Just (tdiffadopted, adopted@(nid,tstmp,ntx,bsz)) -> do
+        let chaints = nodes `zip` map (\n -> HM.lookup (n,sn) hmch) nodes
+        let ochaints = sortOn snd chaints
+        tdiffusion <- forM ochaints $ \(n,ts) -> reconstruct_chain sn adopted n ts
+        pure ((sn,formatTS tstmp,nid,tdiffadopted,ntx,bsz), tdiffusion)
+    Nothing -> pure ((0, "", -1, -1, 0, 0), [])
+  let tdiffusion = catMaybes tdiffusion0
+  reconstruct nodes sns (hmch, hmad, hmld) (tdiffadopted : acc1) (tdiffusion ++ acc2)
+
+reconstruct_leader :: SlotNum -> Maybe ValLeader -> IO ValLeader
+reconstruct_leader _ Nothing = pure (-1, time0)
+reconstruct_leader _ (Just v) = return v
+
+reconstruct_adopt :: SlotNum -> ValLeader -> Maybe ValAdopted
```

Comment:
I see the different style of naming: `processCsv` and `reconstruct_adopt`. It would be better if we had one style:

1. `processCsv` and `reconstructAdopt`
2. `process_csv` and `reconstruct_adopt`